### ISSUE TITLE
Using Reactant-compiled functions together with libraries with Enzyme integration

### ISSUE THREAD

**Issue by nardi** [EnzymeAD/Reactant.jl#1643]
Hi there! This is not really an issue but mostly me asking for help, so if that's not appropriate I'm sorry!

I have a library of JAX functions in Python and I'm trying to use Reactant to embed them in a Julia program. I found that there is a good integration with JAX, following [this test](https://github.com/EnzymeAD/Reactant.jl/blob/main/test/integration/python.jl):

```julia
using PythonCall
import Reactant

jax = pyimport("jax")

x = Float32[1, 2, 3]
y = Reactant.to_rarray(x)

jsum = Reactant.@compile jax.numpy.sum(y)

jsum(y)
```

Then I also saw the integration with Enzyme is straightforward, following some of the code [in here](https://github.com/EnzymeAD/Reactant.jl/blob/main/test/autodiff.jl): 

```julia
using Enzyme

jsum_grad = @Reactant.compile (v -> gradient(Reverse, jax.numpy.sum, v))(y)

jsum_grad(y)
```

according to which the entire differentiation should be compiled, and indeed placing the `@compile` inside `gradient` doesn't work. This makes sense to me, coming from a JAX perspective: the differentiation happens inside the traced function.

However, I can't seem to figure out how to use this integration in a different context. For example, I would like to apply a rootfinding algorithm from NonlinearSolve to one of these functions. Taking the simplest example from [the documentation](https://docs.sciml.ai/NonlinearSolve/v4.7/tutorials/getting_started/#Problem-Type-1:-Solving-Nonlinear-Systems-of-Equations):

```julia
using NonlinearSolve

f(u, p) = u .* u .- p
u0 = [1.0, 1.0]
p = 2.0
prob = NonlinearProblem(f, u0, p)
sol = solve(prob)
```

I tried to replace `f` by a JAX function and the arguments by RArrays. I also thought it would be good to explicitly ask the solver to use Enzyme for autodiff. However, this doesn't work:

```julia
using NonlinearSolve, PythonCall
import Reactant

@pyexec """
def f(u, p):
    return u * u - p
""" => f

u0 = Reactant.to_rarray(Float32[1.0, 1.0])
p = Reactant.to_rarray(Float32[2.0])
solver = NewtonRaphson(autodiff=AutoEnzyme())

# This works:
Reactant.@jit f(u0, p)

# This doesn't work:
sol = Reactant.@jit ((u0, p) -> solve(NonlinearProblem(f, u0, p), solver))(u0, p)
```

That kind of makes sense to me, since the NonlinearSolve package likely performs operations that cannot be traced by Reactant. However, since the function `f` is called and differentiated during the `solve` call, that call must be wrapped in an compilation environment.

Am I missing a different way to use Reactant here? Is there for example a way to trace `f` and pass the traced function to `solve` in a way that it can be executed like a regular function but also still be differentiated with Enzyme? Or do I just have to create all relevant derivative functions (`f_jvp`, `f_vjp`, `f_hvp` etc) myself and pass those to the solver explicitly?

**Comment by wsmoses** [comment]
cc @avik-pal @ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
We need to add a ReactantCore dependency and add @trace to the main solver loop. Easy to handle but I haven't gotten to it because I wasn't able to test on windows before... But now hey you got that so it is time for me to do my part.
